collector/tasks.go

```diff
@@ -0,0 +1,174 @@
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package collector
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"path"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type taskByAction struct {
+	Type   prometheus.ValueType
+	Desc   *prometheus.Desc
+	Value  func(action string, count int64) float64
+	Labels func(action string, count int64) []string
+}
+
+var (
+	taskLabels = []string{"cluster", "action"}
+)
+
+// Task Information Struct
+type Task struct {
+	logger  log.Logger
+	client  *http.Client
+	url     *url.URL
+	actions string
+
+	up                              prometheus.Gauge
+	totalScrapes, jsonParseFailures prometheus.Counter
+
+	byActionMetrics []*taskByAction
+}
+
+// NewTask defines Task Prometheus metrics
+func NewTask(logger log.Logger, client *http.Client, url *url.URL, actions string) *Task {
+	return &Task{
+		logger:  logger,
+		client:  client,
+		url:     url,
+		actions: actions,
+
+		up: prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: prometheus.BuildFQName(namespace, "task_stats", "up"),
+			Help: "Was the last scrape of the ElasticSearch Task endpoint successful.",
+		}),
+		totalScrapes: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: prometheus.BuildFQName(namespace, "task_stats", "total_scrapes"),
+			Help: "Current total Elasticsearch snapshots scrapes.",
+		}),
+		jsonParseFailures: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: prometheus.BuildFQName(namespace, "task_stats", "json_parse_failures"),
+			Help: "Number of errors while parsing JSON.",
+		}),
+		byActionMetrics: []*taskByAction{
+			{
+				Type: prometheus.GaugeValue,
+				Desc: prometheus.NewDesc(
+					prometheus.BuildFQName(namespace, "task_stats", "action_total"),
+					"Number of tasks of a certain action",
+					[]string{"action"}, nil,
+				),
+				Value: func(action string, count int64) float64 {
+					return float64(count)
+				},
+				Labels: func(action string, count int64) []string {
+					return []string{action}
+				},
+			},
+		},
+	}
+}
+
+// Describe adds Task metrics descriptions
+func (t *Task) Describe(ch chan<- *prometheus.Desc) {
+	for _, metric := range t.byActionMetrics {
+		ch <- metric.Desc
+	}
+
+	ch <- t.up.Desc()
+	ch <- t.totalScrapes.Desc()
+	ch <- t.jsonParseFailures.Desc()
+}
+
+func (t *Task) fetchAndDecodeAndAggregateTaskStats() (*AggregatedTaskStats, error) {
+	u := *t.url
+	u.Path = path.Join(u.Path, "/_tasks")
+	u.RawQuery = "group_by=none&actions=" + t.actions
```

Comment:
I'm not sure it's a good idea to use string concatenation here. There is no encoding on t.actions. 

I think we should either use a `URL.Params` and then call `u.RawQuery = params.Encode()` or wrap t.actions in an Encode().

Reply:
Fixed (went the former solution)